File: app/src/main/java/net/xzos/UpgradeAll/server/app/manager/AppManager.java
package net.xzos.UpgradeAll.server.app.manager;

import net.xzos.UpgradeAll.database.RepoDatabase;
import net.xzos.UpgradeAll.server.ServerContainer;
import net.xzos.UpgradeAll.server.app.manager.api.App;
import net.xzos.UpgradeAll.server.app.manager.api.Updater;
import net.xzos.UpgradeAll.server.log.LogUtil;

import org.json.JSONException;
import org.json.JSONObject;
import org.litepal.LitePal;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public class AppManager {

    private LogUtil Log = ServerContainer.AppServer.getLog();
    private static final String TAG = "AppManager";
    private static final String[] LogObjectTag = {"Core", TAG};
    // TODO: 对 Core 日志页做同一父类

    private JSONObject appJson = new JSONObject(); // 存储 Updater Engine 数据

    public void refreshAll(boolean isAuto) {
        if (getAppList().isEmpty())
            initApp();
        List<App> appList = getAppList();
        for (App app : appList) {
            Updater updater = app.getUpdater();
            updater.renew(isAuto);
        }
    }

    private void initApp() {
        List<RepoDatabase> repoDatabase = LitePal.findAll(RepoDatabase.class);
        for (RepoDatabase updateItem : repoDatabase) {
            int appDatabaseId = updateItem.getId();
            setApp(appDatabaseId);
        }
    }

    public App getApp(int appDatabaseId) {
        App app = null;
        try {
            app = (App) appJson.get(String.valueOf(appDatabaseId));
        } catch (JSONException e) {
            Log.e(LogObjectTag, TAG, String.format("getEngine: updateJsonData缺少 %s 项", appDatabaseId));
        }
        return app;
    }

    public void setApp(int appDatabaseId) {
        try {
            appJson.put(String.valueOf(appDatabaseId), new App(appDatabaseId));
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    public void delApp(int appDatabaseId) {
        // TODO: 自initApp维护，数据来源：独立 UI 数据
        appJson.remove(String.valueOf(appDatabaseId));
    }


    private List<App> getAppList() {
        ArrayList<App> appArrayList = new ArrayList<>();
        Iterator<String> keys = appJson.keys();
        while (keys.hasNext()) {
            String key = keys.next();
            appArrayList.add(getApp(Integer.parseInt(key)));
        }
        return appArrayList;
    }
}

File: app/src/main/kotlin/net/xzos/upgradeAll/server/app/manager/AppManager.kt
package net.xzos.upgradeAll.server.app.manager

import net.xzos.upgradeAll.data.database.litepal.RepoDatabase
import net.xzos.upgradeAll.data.database.manager.AppDatabaseManager
import net.xzos.upgradeAll.server.app.manager.module.App

internal object AppManager {

    private val appSet = hashSetOf<App>() // 存储 Updater Engine 数据

    init {
        initApp()
    }

    private fun initApp() {
        val appDatabase = AppDatabaseManager.appDatabases
        for (appItem in appDatabase) {
            setApp(appItem)
        }
    }

    fun getApps(): HashSet<App> =
            appSet

    fun getApp(databaseId: Int? = null, uuid: String? = null): App? {
        for (app in appSet) {
            if ((uuid != null && app.appDatabase.extraData?.cloudAppConfig?.uuid == uuid)
                    || (databaseId != null && app.appDatabase.id == databaseId))
                return app
        }
        return null
    }

    fun addApp(app: App) {
        appSet.add(app)
    }

    fun setApp(appDatabase: RepoDatabase) {
        appSet.add(App(appDatabase))
    }

    fun delApp(app: App) {
        // TODO: initApp自维护，数据来源：独立 UI 数据
        appSet.remove(app)
    }

    fun renewAppInHub(hubUuid: String) {
        for (app in appSet) {
            if (app.appDatabase.api_uuid == hubUuid) {
                app.renewEngine()
            }
        }
    }
}
